### ISSUE TITLE
Bug when adding an unknown value in a ReferenceList

### ISSUE THREAD

**Issue by yohanboniface** [gristlabs/grist-core#341]
HI,

In a ReferenceList column, I add an unknown value (thus not existing in the reference table), but I then click on Enter, I do not click on the "+":

[Screencast from 2022-11-02 20-08-08.webm](https://user-images.githubusercontent.com/146023/199580157-00a15cf4-984a-4758-af23-80da5859661d.webm)

I'm expecting the unknown value to be added with the red box, as in the preview.

Thanks for looking at it :)

Yohan

c @LouisDelbosc @arnaudpeich

**Comment by paulfitz** [comment]
Makes sense as an improvement. @georgegevoian do you have insight into how big a project this might be?

**Comment by georgegevoian** [comment]
Makes sense to me too.

If I recall correctly, the client flags cells as being invalid if the type of the value received from the server isn't the type the widget expects. In this case, it's because the server is returning a string representation of the reference(s), while the client expects a Grist-encoded tuple (e.g. `["L", 1, 2, "something else"]`). So the first step would be to update the backend to send a tuple.

Once that's done, the rest is just updating the client to pick out which references in the list - if any - are invalid, and to format them so that they have a red outline. We've done this already for Choice List cells, and much of it should be re-usable for Reference List cells.

One other upside to this change is that opening the editor for a cell with invalid references should no longer clear the editor input.

I don't think it's a large project @paulfitz. The only part I'm unsure about is updating the data engine to send a tuple instead of a string - it seems like a simple change, but there may be other parts of the system that need to be updated alongside it. The changes to the client should be minimal and straightforward.

**Comment by dsagal** [comment]
Currently, an invalid value in Grist is represented in the data engine by an `AltText` object, which is just a wrapper for a string (in DB and in the client, it _is_ just a string). The root of the problem is that a ReferenceList value with invalid members is an invalid value, but is entirely unhelpful when represented as AltText.

The tricky part I anticipate is that we'll actually want to treat the value as mostly-valid, and only treat the invalid parts as invalid. I.e. it's not just about showing it in cells; it would feel wrong if the entire value is treat as invalid, e.g. omitted from summary tables, or filters, etc. Representing it as `["L", 1, 2, "something else"]` is reasonable, and it will look valid to most code. But then anything that handles ReferenceLists will need to handle invalid parts inside them. This applies both in the client (e.g. widget-linking and filtering), and in the data engine (e.g. `$refList[0]` or `$refList.Field` or summary tables grouped by a RefList column). So @georgegevoian is right to worry that there are other parts of the system that will need updating.

**Comment by hexaltation** [comment]
After some testing in our instance, it seems to be fixed.
Feel free to reopen if needed.
